### DISCUSSION TITLE
"TS parser failed" when opening large files

### DISCUSSION THREAD

**Discussion by larsks** [helix-editor/helix#12263] [General]
I am using helix to open a large (22MB) JSON file. The open file does not receive any syntax highlighting, and helix reports (in `helix.log`):

```
2024-12-15T12:28:35.974 helix_core::syntax [ERROR] TS parser failed, disabling
TS for the current buffer: Err(Cancelled)
```

Opening smaller files works without a problem. The file is valid JSON (as confirmed with e.g. `jq`, `json_verify` from [yajl], `python -m json.tool`).

[yajl]: https://lloyd.github.io/yajl/

**Comment by kirawi** [comment]
This is intentional to avoid Helix being unusable when syntax highlighting takes too long.

**Comment by larsks** [comment]
What determines the threshold for whether or not syntax highlighting is disabled? Is that adjustable? Is there a way to get some sort of notification when this happens?

Looking at other editors...

- `neovim` takes a couple of seconds to open the file, but does successfully open and highlight the file. It displays the message "Big file detected. Some neovim features have been disabled."
- `vis` opens the file with no noticeable delay and highlights it correctly.
- `kakoune` opens the file with no noticeable delay and highlights it correctly.

It looks like `helix` is the only editor that actually refused to highlight the file.

**Reply by TornaxO7** [reply]
> What determines the threshold for whether or not syntax highlighting is disabled?

[A simple timeout](https://github.com/helix-editor/helix/blob/master/helix-core/src/syntax.rs#L1254).

> Is that adjustable?

Looks like not, but I'm not really familiar with the codebase.

> Is there a way to get some sort of notification when this happens?

AFAIK:

![peepoNo](https://github.com/user-attachments/assets/7592c415-7e2b-44d2-8720-8bd464806820)

**Reply by kirawi** [reply]
Tree-sitter has much higher time and space complexity compared to regex AFAIK
